worker/src/RTC/RtpStreamSend.cpp

```diff
@@ -280,118 +417,93 @@ namespace RTC
+			auto clockRate{ this->params.clockRate };
+
+			// Go through all buffer items starting with the first and free all storage
+			// items that contain packets older than `MaxRetransmissionDelay`.
+			for (uint32_t i{ 0 }; i <= static_cast<uint32_t>(MaxSeq); ++i)
+			{
+				auto* checkedStorageItem = this->storageItemBuffer.GetFirst();
 
-			// Reset the first storage item.
-			resetStorageItem(firstStorageItem);
+				// Packets can arrive out of order, in which case we'll miss some storage items.
+				if (checkedStorageItem)
+				{
+					// This is the storage item we have just inserted, no need to go further.
+					if (!checkedStorageItem->originalPacket)
+						break;
 
-			// Unfill the buffer start item.
-			this->buffer[this->bufferStartIdx] = nullptr;
+					auto checkedPacketTs{ checkedStorageItem->originalPacket->GetTimestamp() };
+					auto diffTs{ packetTs - checkedPacketTs };
 
-			// Move the buffer start index.
-			UpdateBufferStartIdx();
+					// Account for wrapping around.
+					if (diffTs > MaxTs / 2)
+					{
+						diffTs = MaxTs - diffTs;
+					}
 
-			// Take the freed storage item.
-			storageItem       = firstStorageItem;
-			this->buffer[seq] = storageItem;
+					// Cleanup is finished if we found an item with recent enough packet, but also account
+					// for out-of-order packets.
+					if (
+					  static_cast<uint32_t>(diffTs * 1000 / clockRate) < this->retransmissionBufferSize ||
+					  RTC::SeqManager<uint32_t>::IsSeqLowerThan(packetTs, checkedPacketTs))
+						break;
+
+					// Reset (free RTP packet) the old storage item.
+					resetStorageItem(checkedStorageItem);
+					// Return into the pool.
+					StorageItemPool.Return(checkedStorageItem);
+					// Unfill the buffer start item.
+					MS_ASSERT(this->storageItemBuffer.RemoveFirst(), "Storage item must be used");
+				}
+			}
 		}
 
-		// Clone the packet into the retrieved storage item.
-		storageItem->packet = packet->Clone(storageItem->store);
-	}
-
-	void RtpStreamSend::ClearBuffer()
-	{
-		MS_TRACE();
-
-		if (this->storage.empty())
-			return;
-
-		for (uint32_t idx{ 0 }; idx < this->buffer.size(); ++idx)
+		// Only clone once and only if necessary.
+		if (!clonedPacket.get())
 		{
-			auto* storageItem = this->buffer[idx];
-
-			if (!storageItem)
-			{
-				MS_ASSERT(!this->buffer[idx], "key should be NULL");
-
-				continue;
-			}
-
-			// Reset (free RTP packet) the storage item.
-			resetStorageItem(storageItem);
+			auto clone = packet->Clone();
 
-			// Unfill the buffer item.
-			this->buffer[idx] = nullptr;
+			// Move the RtpPacket pointer into clonedPacket shared pointer.
+			clonedPacket.swap(clone);
 		}
 
-		// Reset buffer.
-		this->bufferStartIdx = 0;
-		this->bufferSize     = 0;
+		// Store original packet and some extra info into the retrieved storage item.
+		storageItem->originalPacket = clonedPacket;
+		storageItem->ssrc           = clonedPacket->GetSsrc();
+		storageItem->sequenceNumber = clonedPacket->GetSequenceNumber();
```

Comment:
This is **the** or at least **a** culprit. We are setting here the SSRC of a different `RtpStreamSend`.., the one that cloned the packet. We need to set the item ssrc to `packet->GetSsrc()`, same with sequence number.

I'll add the changes along with some other sanitizations in a PR.

cc: @nazar-pc @Hartigan

Reply:
@jmillan ```timestamp``` is needed too?

Reply:
yes

Reply:
Timestamp wasn't updated before and doesn't need to be updated now, it is just copied from source packets and there is no problem in that.

SSRC and sequence number are also fine, they are changed to correct values before packet gets here in corresponding consumers, for instance: https://github.com/versatica/mediasoup/blob/af3b681f1355835931c9d4a3d960cbe0a42e739f/worker/src/RTC/PipeConsumer.cpp#L238-L240

So I don't believe there is any error here, cloned packet is updated before it is processed, since it can be later modified again we store SSRC and sequence number in storage item (they weren't stored before because each storage item owned exclusive copy of the packet).

Reply:
The problem here was that `storageItem->ssrc` was being set with the `clonedPacket's` one instead of with the original packet.

That is handled in the sanitisation branch.